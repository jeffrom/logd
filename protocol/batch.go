package protocol

import (
	"bufio"
	"bytes"
	"fmt"
	"hash/crc32"
	"io"

	"github.com/jeffrom/logd/config"
	"github.com/pkg/errors"
)

// MaxTopicSize represents the maximum topic length. It is 255, as that is the
// maximum directory length in linux.
const MaxTopicSize = 255

// Batch represents a collection of Messages
// BATCH <size> <topic> <checksum> <messages>\r\n<data>
// NOTE no trailing newline after the data
type Batch struct {
	conf     *config.Config
	Size     uint64
	Checksum uint32
	Messages int
	topic    []byte
	ntopic   int
	msgs     []*Message
	body     []byte
	digitbuf [32]byte
	msgBuf   *bytes.Buffer
	firstOff uint64
	wasRead  bool
}

// NewBatch returns a new instance of a batch
func NewBatch(conf *config.Config) *Batch {
	b := &Batch{
		conf:   conf,
		msgBuf: &bytes.Buffer{},
		topic:  make([]byte, MaxTopicSize),
		msgs:   make([]*Message, 1000),
	}

	return b
}

func (b *Batch) String() string {
	return fmt.Sprintf("Batch<%p Messages: %d, Size: %d>", b, b.Messages, b.Size)
}

// Reset puts a batch in an initial state so it can be reused
func (b *Batch) Reset() {
	b.Size = 0
	b.Checksum = 0
	b.Messages = 0
	b.ntopic = 0
	b.firstOff = 0
	b.wasRead = false
	b.msgBuf.Reset()
}

// Empty returns true if the batch contains no messages
func (b *Batch) Empty() bool {
	if b.Messages <= 0 {
		return true
	}
	return false
}

func (b *Batch) ensureBuf() {
	if b.body == nil {
		b.body = make([]byte, b.conf.MaxBatchSize)
	}
}

// SetTopic sets the topic for a batch.
func (b *Batch) SetTopic(topic []byte) {
	copy(b.topic, topic)
	b.ntopic = len(topic)
}

// Topic returns the topic for the batch.
func (b *Batch) Topic() string {
	return string(b.TopicSlice())
}

// TopicSlice returns the topic for the batch as a byte slice. The byte slice
// is not copied.
func (b *Batch) TopicSlice() []byte {
	return b.topic[:b.ntopic]
}

// FromRequest parses a request, populating the batch. If validation fails, an
// error is returned.
func (b *Batch) FromRequest(req *Request) (*Batch, error) {
	if req.nargs != argLens[CmdBatch] {
		return b, errInvalidNumArgs
	}

	// fmt.Printf("%s(%q, %q)\n", &req.Name, req.args[0], req.args[1])
	n, err := asciiToUint(req.args[0])
	if err != nil {
		return b, err
	}
	b.Size = n

	b.SetTopic(req.args[1])

	n, err = asciiToUint(req.args[2])
	if err != nil {
		return b, err
	}
	b.Checksum = uint32(n)

	n, err = asciiToUint(req.args[3])
	if err != nil {
		return b, err
	}
	b.Messages = int(n)

	b.body = req.body[:req.bodysize]

	b.firstOff = uint64(len(req.envelope) + termLen)
	return b, b.Validate()
}

// Validate checks the batch's checksum
// TODO should add config.MaxMessageSize and config.MaxMessagesPerBatch and
// check them here, maybe?
func (b *Batch) Validate() error {
	// if size > MaxBatchSize || crc doesn't match
	if b.Size > uint64(b.conf.MaxBatchSize) {
		return errors.New("batch too large")
	}
	if b.Checksum != b.calculateChecksum() {
		return errors.New("checksum didn't match")
	}
	return nil
}

// Bytes returns a slice of raw bytes. Used by EventQ to write directly to the
// log.
func (b *Batch) Bytes() []byte {
	if b.body == nil {
		return nil
	}
	return b.body[:b.Size]
}

// Append adds a new message's bytes to the batch
func (b *Batch) Append(p []byte) error {
	if b.Messages > len(b.msgs)-1 {
		msgs := make([]*Message, len(b.msgs)*2)
		copy(msgs, b.msgs)
		b.msgs = msgs
	}
	if b.msgs[b.Messages] == nil {
		b.msgs[b.Messages] = NewMessage(b.conf)
	}
	msg := b.msgs[b.Messages]
	msg.Reset()
	msg.Body = p
	msg.Size = len(p)

	b.Messages++
	b.Size += uint64(msg.calcSize())
	return nil
}

// AppendMessage adds a new message to the batch
func (b *Batch) AppendMessage(m *Message) error {
	b.msgs[b.Messages] = m
	b.Messages++
	return nil
}

// SetMessages sets all messages in a batch
func (b *Batch) SetMessages(msgs []*Message) {
	b.msgs = msgs
	b.Messages = len(msgs)
}

// MessageBytes returns a byte slice of the batch of messages.
func (b *Batch) MessageBytes() []byte {
	return b.body[:b.Size]
}

// Annotate adds firstOffset and offsetDelta information to each message in the
// batch
func (b *Batch) Annotate() error {
	b.firstOff = b.calculateFirstOffset()

	var n uint64
	for i := 0; i < b.Messages; i++ {
		m := b.msgs[i]
		m.firstOffset = b.firstOff
		m.offsetDelta = b.firstOff + n
		msgSize, err := b.messageFullSize(m)
		if err != nil {
			return err
		}
		n += uint64(msgSize)
	}

	return nil
}

// SetChecksum sets the batch's crc32
func (b *Batch) SetChecksum() {
	b.Checksum = b.calculateChecksum()
}

// CalcSize calculates the full byte size of the batch. this is intended to be
// called to make sure the batch isn't larger than config.MaxBatchSize, so
// we're assuming the crc is the longest possible uint32 for now to save
// calculating the crc for every message
func (b *Batch) CalcSize() int {
	l := len(bbatchStart)       // `BATCH `
	l += asciiSize(int(b.Size)) // <size>
	l += len(bspace)            // ` `
	l += b.ntopic               // `<topic>`
	l += len(bspace)            // ` `
	l += maxCRCSize             // <crc>
	l += len(bspace)            // ` `
	l += asciiSize(b.Messages)  // <messages>
	l += termLen                // `\r\n`
	l += int(b.Size)            // <data>
	return l
}

func (b *Batch) calculateChecksum() uint32 {
	return crc32.Checksum(b.Bytes(), crcTable)
}

func (b *Batch) buildBodyBytes() error {
	b.msgBuf.Reset()
	for i := 0; i < b.Messages; i++ {
		m := b.msgs[i]
		n, err := m.WriteTo(b.msgBuf)
		if err != nil {
			return err
		}

		m.fullSize = int(n)
	}

	l := b.msgBuf.Len()
	b.Size = uint64(l)
	b.ensureBuf()
	// fmt.Printf("buildBodyBytes: %q\n", b.msgBuf.Bytes())
	// fmt.Println(len(b.body), b.msgBuf.Len())
	if b.msgBuf.Len() > len(b.body) {
		return errTooLarge
	}
	copy(b.body[:b.Size], b.msgBuf.Bytes())

	return nil
}

func (b *Batch) messageFullSize(m *Message) (int, error) {
	if m.fullSize > 0 {
		return m.fullSize, nil
	}

	b.msgBuf.Reset()
	n, err := m.WriteTo(b.msgBuf)
	m.fullSize = int(n)
	return m.fullSize, err
}

// WriteTo implements io.WriterTo.
func (b *Batch) WriteTo(w io.Writer) (int64, error) {
	if !b.wasRead {
		if err := b.buildBodyBytes(); err != nil {
			return 0, err
		}

		b.SetChecksum()
	}

	var total int64
	n, err := w.Write(bbatchStart)
	total += int64(n)
	if err != nil {
		return total, err
	}

	l := uintToASCII(uint64(b.Size), &b.digitbuf)
	n, err = w.Write(b.digitbuf[l:])
	total += int64(n)
	if err != nil {
		return total, err
	}

	n, err = w.Write(bspace)
	total += int64(n)
	if err != nil {
		return total, err
	}

	n, err = w.Write(b.TopicSlice())
	total += int64(n)
	if err != nil {
		return total, err
	}

	n, err = w.Write(bspace)
	total += int64(n)
	if err != nil {
		return total, err
	}

	l = uintToASCII(uint64(b.Checksum), &b.digitbuf)
	n, err = w.Write(b.digitbuf[l:])
	total += int64(n)
	if err != nil {
		return total, err
	}

	n, err = w.Write(bspace)
	total += int64(n)
	if err != nil {
		return total, err
	}

	l = uintToASCII(uint64(b.Messages), &b.digitbuf)
	n, err = w.Write(b.digitbuf[l:])
	total += int64(n)
	if err != nil {
		return total, err
	}

	n, err = w.Write(bnewLine)
	total += int64(n)
	if err != nil {
		return total, err
	}

	n, err = w.Write(b.MessageBytes())
	total += int64(n)
	if err != nil {
		return total, err
	}

	return total, err
}

// ReadFrom implements io.ReaderFrom
func (b *Batch) ReadFrom(r io.Reader) (int64, error) {
	n, err := b.readFromBuf(r.(*bufio.Reader))
	if err != nil {
		return n, err
	}
	b.wasRead = true
	return n, err
}

// FirstOffset returns the offset delta of the first message
func (b *Batch) FirstOffset() uint64 {
	if b.firstOff == 0 {
		b.firstOff = b.calculateFirstOffset()
	}
	return b.firstOff
}

// readFromBuf reads a batch from a *bufio.Reader
func (b *Batch) readFromBuf(r *bufio.Reader) (int64, error) {
	var total int64
	total, err := b.readEnvelope(r)
	if err != nil {
		return total, err
	}

	n, err := b.readData(r)
	total += n
	return total, err
}

// readEnvelope reads the batch protocol envelope
func (b *Batch) readEnvelope(r *bufio.Reader) (int64, error) {
	var total int64
	word, err := r.ReadSlice(' ')
	total += int64(len(word))

	// fmt.Printf("%q %+v\n", word, err)
	if len(word) == 0 {
		if err == io.EOF {
			return total, err
		}
		return total, errInvalidProtocolLine
	}
	if !bytes.Equal(word, bbatchStart) {
		return total, errors.Wrap(errInvalidProtocolLine, "batch envelope didn't start with BATCH")
	}

	if err != nil {
		return total, err
	}

	word, err = r.ReadSlice(' ')
	total += int64(len(word))
	if err != nil {
		return total, err
	}

	var n uint64
	n, err = asciiToUint(word[:len(word)-1])
	if err != nil {
		return total, err
	}
	b.Size = n

	word, err = r.ReadSlice(' ')
	total += int64(len(word))
	if err != nil {
		return total, err
	}

	b.SetTopic(word[:len(word)-1])

	word, err = r.ReadSlice(' ')
	total += int64(len(word))
	if err != nil {
		return total, err
	}

	n, err = asciiToUint(word[:len(word)-1])
	if err != nil {
		return total, err
	}
	b.Checksum = uint32(n)

	word, err = r.ReadSlice('\n')
	total += int64(len(word))
	if err != nil {
		return total, err
	}
	n, err = asciiToUint(word[:len(word)-termLen])
	if err != nil {
		return total, err
	}
	b.Messages = int(n)
	return total, err
}

// readData reads the data portion of a batch command.
func (b *Batch) readData(r *bufio.Reader) (int64, error) {
	var total int64

	// TODO this check is redundant, should check the total batch size above this
	if b.Size > uint64(b.conf.MaxBatchSize) {
		return total, errTooLarge
	}
	b.ensureBuf()
	n, err := io.ReadFull(r, b.body[:b.Size])
	total += int64(n)

	return total, err
}

func (b *Batch) calculateFirstOffset() uint64 {
	n := uint64(len(bbatchStart) +
		uintToASCII(b.Size, &b.digitbuf) +
		len(bspace) +
		uintToASCII(uint64(b.Messages), &b.digitbuf) +
		termLen +
		b.ntopic)

	return n
}

// Copy returns a newly allocated copy
func (b *Batch) Copy() *Batch {
	batch := NewBatch(b.conf)
	batch.Size = b.Size
	batch.Checksum = b.Checksum
	batch.Messages = b.Messages
	batch.SetTopic(b.TopicSlice())
	batch.body = make([]byte, len(b.body))
	copy(batch.body, b.body)
	batch.wasRead = true
	return batch
}
