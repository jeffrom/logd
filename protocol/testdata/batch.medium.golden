BATCH 4472 default 3621047634 3
MSG 534
When Marx undertook his critique of the capitalistic mode of production, this mode was in its infancy. Marx directed his efforts in such a way as to give them prognostic value. He went back to the basic conditions underlying capitalistic production and through his presentation showed what could be expected of capitalism in the future. The result was that one could expect it not only to exploit the proletariat with increasing intensity, but ultimately to create conditions which would make it possible to abolish capitalism itself.
MSG 1287
The transformation of the superstructure, which takes place far more slowly than that of the substructure, has taken more than half a century to manifest in all areas of culture the change in the conditions of production. Only today can it be indicated what form this has taken. Certain prognostic requirements should be met by these statements. However, theses about the art of the proletariat after its assumption of power or about the art of a classless society would have less bearing on these demands than theses about the developmental tendencies of art under present conditions of production. Their dialectic is no less noticeable in the superstructure than in the economy. It would therefore be wrong to underestimate the value of such theses as a weapon. They brush aside a number of outmoded concepts, such as creativity and genius, eternal value and mystery – concepts whose uncontrolled (and at present almost uncontrollable) application would lead to a processing of data in the Fascist sense. The concepts which are introduced into the theory of art in what follows differ from the more familiar terms in that they are completely useless for the purposes of Fascism. They are, on the other hand, useful for the formulation of revolutionary demands in the politics of art.
MSG 2616
In principle a work of art has always been reproducible. Man-made artifacts could always be imitated by men. Replicas were made by pupils in practice of their craft, by masters for diffusing their works, and, finally, by third parties in the pursuit of gain. Mechanical reproduction of a work of art, however, represents something new. Historically, it advanced intermittently and in leaps at long intervals, but with accelerated intensity. The Greeks knew only two procedures of technically reproducing works of art: founding and stamping. Bronzes, terra cottas, and coins were the only art works which they could produce in quantity. All others were unique and could not be mechanically reproduced. With the woodcut graphic art became mechanically reproducible for the first time, long before script became reproducible by print. The enormous changes which printing, the mechanical reproduction of writing, has brought about in literature are a familiar story. However, within the phenomenon which we are here examining from the perspective of world history, print is merely a special, though particularly important, case. During the Middle Ages engraving and etching were added to the woodcut; at the beginning of the nineteenth century lithography made its appearance. With lithography the technique of reproduction reached an essentially new stage. This much more direct process was distinguished by the tracing of the design on a stone rather than its incision on a block of wood or its etching on a copperplate and permitted graphic art for the first time to put its products on the market, not only in large numbers as hitherto, but also in daily changing forms. Lithography enabled graphic art to illustrate everyday life, and it began to keep pace with printing. But only a few decades after its invention, lithography was surpassed by photography. For the first time in the process of pictorial reproduction, photography freed the hand of the most important artistic functions which henceforth devolved only upon the eye looking into a lens. Since the eye perceives more swiftly than the hand can draw, the process of pictorial reproduction was accelerated so enormously that it could keep pace with speech. A film operator shooting a scene in the studio captures the images at the speed of an actor’s speech. Just as lithography virtually implied the illustrated newspaper, so did photography foreshadow the sound film. The technical reproduction of sound was tackled at the end of the last century. These convergent endeavors made predictable a situation which Paul Valery pointed up in this sentence:
